morpho5/modules/meshslice.morpho

```diff
@@ -0,0 +1,211 @@
+/* Meshslice - Slices a Mesh along a plane */
+
+import meshtools
+import delaunay
+
+var _errSlcEmpty = Error("SlcEmpty", "No slice has yet been taken.")
+
+class Intersection {
+  init(u, x) {
+    self.u = u
+    self.x = x
+  }
+}
+
+class MeshSlicer {
+  init(mesh, ptol=1e-8) {
+    self.mesh = mesh
+    self.newmesh = nil 
+    self.pt = nil
+    self.dirn = nil
+    self.basis = nil 
+    self.ptol = ptol
+    self.pairs = nil
+    self.vert = nil
+    self.mb = nil
+  }
+
+  basis() { // Creates a basis perpendicular to the dirn 
+    var d = []
+    for (x in self.dirn) d.append(abs(x))
+    var order = d.order()
+
+    var e = [self.dirn]
+    for (i in 0..1) {
+      var ei = Matrix(self.dirn.count()) // Build a vector in that direction 
+      ei[order[i]]=1
+      e.append(ei)
+    }
+
+    var f = self.orthogonalize(e) 
+    self.basis = [ f[1], f[2] ]
+  }
+
+  orthogonalize(v) {  
+    var u = []
+    for (i in 0...v.count()) {
+      var ui = v[i]
+      for (uj in u) ui -= uj*uj.inner(v[i])/uj.inner(uj) 
+      u.append(ui/ui.norm())
+    }
+    return u
+  } 
+
+  setpt(pt, dirn) {
+    self.pt = nil
+    self.dirn = nil
+    if (islist(pt)) self.pt = Matrix(pt)
+    if (islist(dirn)) self.dirn = Matrix(dirn)
+    self.basis()
+  }
+
+  testintersection(indices) { // Finds if an element with indices intersects with the plane
+    var minus=false, plus=false
+    for (i in indices) {
+      var x = self.vert.column(i)
+      var t = sign((x - self.pt).inner(self.dirn))
+      if (t>0) plus=true
+      if (t<0) minus=true
+    }
+    return (plus && minus)
+  }
+
+  findintersection(i,j) { // returns the point between i,j that lies on the intersection plane
+    var x0 = self.vert.column(i)
+    var x1 = self.vert.column(j)
+
+    var t0 = (x0 - self.pt).inner(self.dirn)
+    var t1 = (x1 - self.pt).inner(self.dirn)
+    if (t0*t1>0) return nil // check if the points lie on the same side of the plane
+
+    var u = t0/(t0-t1)
+
+    return Intersection(u, (1-u)*x0 + u*x1)
+  }
+
+  vertexpairs(el) { // return all possible pairs of vertices from a list of vertices
+    var out = []
+    var n = el.count()
+    for (i in 0...n-1) for (j in i+1...n) {
```

Comment:
Don't we have tuples for this?

Reply:
Yes that's a good point; this predated tuples!